### PR TITLE
Processor: Transfer Hook

### PR DESCRIPTION
**Pull request by buffalojoec**
This PR introduces the processor for the SPL Transfer Hook interface's
`ExecuteInstruction` instruction.

Just like `HarvestRewards`, this instruction calculates each holder's share
of any unseen total rewards, this time updating their `unharvested_rewards`
and `last_seen_total_rewards` post-transfer.

### CODE REVIEW

**Review by joncinque — COMMENTED** [review]
I haven't looked at the tests, but there were a few important things in the main implementation that I wanted to point out

File: program/src/processor.rs [hunk 1]
```diff
@@ -362,10 +368,194 @@ fn process_harvest_rewards(_program_id: &Pubkey, _accounts: &[AccountInfo]) -> P
...
+    accounts: &[AccountInfo],
     _amount: u64,
 ) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let source_token_account_info = next_account_info(accounts_iter)?;
+    let mint_info = next_account_info(accounts_iter)?;
+    let destination_token_account_info = next_account_info(accounts_iter)?;
+    let _source_owner_info = next_account_info(accounts_iter)?;
+    let _extra_metas_info = next_account_info(accounts_iter)?;
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/src/processor.rs:381]: Nope, neither of these two are used!

File: program/src/processor.rs [hunk 2]
```diff
@@ -362,10 +368,194 @@ fn process_harvest_rewards(_program_id: &Pubkey, _accounts: &[AccountInfo]) -> P
...
+
+        let unseen_unharvested_rewards =
+            calculate_reward_share(token_supply, token_account_balance, unseen_total_rewards)?;
+        holder_rewards_state.unharvested_rewards = holder_rewards_state
+            .unharvested_rewards
+            .saturating_sub(unseen_unharvested_rewards);
+    }
+
+    // Update the destination holder rewards account.
+    {
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/src/processor.rs:487]: Yep, I can. I was considering factoring a few things into functions once the other processors landed. When `HarvestRewards` goes in I'll explore this a bit further.

File: program/src/processor.rs [hunk 3]
```diff
@@ -362,10 +368,194 @@ fn process_harvest_rewards(_program_id: &Pubkey, _accounts: &[AccountInfo]) -> P
...
+                return Err(PaladinRewardsError::TokenAccountMintMismatch.into());
+            }
+
+            // Ensure the provided token account is transferring.
+            let extension = token_account.get_extension::<TransferHookAccount>()?;
+            if !bool::from(extension.transferring) {
+                return Err(TransferHookError::ProgramCalledOutsideOfTransfer.into());
+            }
+
+            token_account.base.amount
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/src/processor.rs:476]: Got it, I'll update it. I misunderstood how this calculation was supposed to work.

**Comment by buffalojoec** [comment]
@joncinque thanks for the review. Since you didn't look at the test yet, I overwrote that commit to clean them up. Afterward I've added a new commit to address your feedback on the calculations, then the helper.

Next commit will be adding loads of test cases...

**Comment by buffalojoec** [comment]
Closing until the reworking of the pool calculations is completed. Then I'll reopen in a new PR, rebased on the new code.

**Pull request closed without merging**